ls1intum/Artemis
# `Development`: Refactor quiz exercise submission

### Checklist
#### General


#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/
#### Client

### Motivation and Context



Most of the implementation of Quiz Submission (used by existing Quiz Exercise) will be reused for Quiz Exam Submission (used for Quiz Questions from Quiz Pool that will be added in a separate PR). To avoid code duplication, some of the implementations of Quiz Submission need to be abstracted.

### Description
1. Created AbstractQuizSubmission model class and moved common properties of QuizSubmission and QuizExamSubmission to this model class
2. Modified SubmittedAnswer model to depend on AbstractQuizSubmission instead of QuizSubmission
3. Created AbstractQuizSubmissionService and moved the common codes of QuizSubmissionService (used by existing Quiz Exercise) and QuizExamSubmissionService (Quiz Questions from Quiz Pool that will be added in a separate PR) to this class
4. Refactor the calculateAndUpdateScores method to receive a  list of QuizQuestion instead of QuizExercise, so that the method can be reused by QuizExamSubmissionService later.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Quiz Exercise (Can be any type: Synchronize, Batched, or Individual)

1. Log in to Artemis as a Student
5. Participate in the Quiz Exercise
6. Verify that the answers are saved by refreshing the page after submitting and ensure that the answers are saved
7. If Quiz is Batched or Individual, click **End Now** as Instructor. Or wait until the time ends for a Synchronized Quiz
8. Verify that the students can see the result
9. As an instructor, verify that the preview mode of the Quiz Exercise is still working by clicking the Preview Quiz button, submitting the answers, and verifying that the results are shown
10. As an instructor, verify that the solution mode of the Quiz Exercise is still working by clicking the Sample Solution button and verifying that the solutions are displayed

#### Exam Mode Testing



Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Quiz Exercise

1. Log in to Artemis as a Student
2. Participate in the exam
3. Click the **Save** button
4. Refresh the page and verify that the answers are saved
5. Submit the exam
11. Log out and log in as an Instructor
12. Navigate to the Exam and click Assessment
13. Wait until the exam is finished and the grace period has ended, then click Evaluate Quiz Exercises
14. Log out and log in as a Student
15. Verify that the students can see the result

### Review Progress






#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage



| Class/File | Branch | Line |
|------------|-------:|-----:|
| AbstractQuizSubmission.java | 80% | 92% |
| QuizSubmission.java | n/a | 100% |
| SubmittedAnswer.java | 50% | 91% |
| AbstractQuizSubmissionService.java | n/a | 88% |
| ExerciseService.java | 81% | 96% |
| QuizExerciseService.java | 85% | 96% |
| QuizSubmissionService.java | 73% | 83% |
| ExamQuizService.java | 76% | 87% |
| QuizScheduleService.java | 72% | 89% |
| ParticipationResource.java | 69% | 91% |
| QuizSubmissionResource.java | 93% | 100% |
| quiz-exam-submission.component.ts | 77.61% | 92.72% |

### Screenshots



**Exercise**
https://user-images.githubusercontent.com/10885503/231222471-1037787a-5ab8-488c-904e-48aaa6d76957.mov

**Exam**
https://user-images.githubusercontent.com/10885503/231222458-affbc09f-7198-49fa-b13f-8de0098bf03c.mov
